CMakeLists.txt

```diff
@@ -639,24 +639,126 @@ hpx_add_compile_flag(-DBOOST_DISABLE_ASSERTS
 
 # Make sure we compile in C++11 mode (MSVC uses it automatically)
 if(NOT MSVC)
-  #set(_cxx11_support FALSE)
-  # Try -std=c++11 first
-  check_cxx_compiler_flag(-std=c++11 HPX_WITH_CXX11)
-  # ... If it failed try -std=c++0x
-  if(NOT HPX_WITH_CXX11)
-    check_cxx_compiler_flag(-std=c++0x HPX_WITH_CXX0X)
-    if(HPX_WITH_CXX0X)
-      set(CXX11_FLAG -std=c++0x)
-      set(HPX_WITH_CXX11 TRUE)
-    endif()
+  # Try -std=c++14 first
+  check_cxx_compiler_flag(-std=c++14 HPX_WITH_CXX14)
+  if(HPX_WITH_CXX14)
+    set(CXX_FLAG -std=c++14)
   else()
-    set(CXX11_FLAG -std=c++11)
-  endif()
-  if(NOT HPX_WITH_CXX11)
-    hpx_error("Unable to compile HPX without C++11 support in your compiler. HPX needs at the very least rvalue references with proper move semantics")
+    # ... otherwise try -std=c++0y
+    check_cxx_compiler_flag(-std=c++0y HPX_WITH_CXX0Y)
+    if(HPX_WITH_CXX0Y)
+      set(CXX_FLAG -std=c++0y)
+    else()
+      # ... otherwise try -std=c++11
+      check_cxx_compiler_flag(-std=c++11 HPX_WITH_CXX11)
+      if(HPX_WITH_CXX11)
+        set(CXX_FLAG -std=c++11)
+      else()
+        # ... otherwise try -std=c++0x
+        check_cxx_compiler_flag(-std=c++0x HPX_WITH_CXX0X)
+        if(HPX_WITH_CXX0X)
+          set(CXX_FLAG -std=c++0x)
+        endif()
+      endif()
+    endif()
   endif()
-  hpx_add_compile_flag(${CXX11_FLAG})
 
+  hpx_add_compile_flag(${CXX_FLAG})
+endif()
+
+################################################################################
+# C++ feature tests
+################################################################################
+
+# Check the availability of certain C++11 language features
+hpx_check_for_cxx11_alias_templates(
+  DEFINITIONS HPX_HAVE_CXX11_ALIAS_TEMPLATES)
+
+hpx_check_for_cxx11_auto(
+  DEFINITIONS HPX_HAVE_CXX11_AUTO)
+
+hpx_check_for_cxx11_constexpr(
+  DEFINITIONS HPX_HAVE_CXX11_CONSTEXPR)
+
+hpx_check_for_cxx11_decltype(
+  DEFINITIONS HPX_HAVE_CXX11_DECLTYPE BOOST_RESULT_OF_USE_DECLTYPE)
+
+hpx_check_for_cxx11_defaulted_functions(
+  DEFINITIONS HPX_HAVE_CXX11_DEFAULTED_FUNCTIONS)
+
+hpx_check_for_cxx11_deleted_functions(
+  DEFINITIONS HPX_HAVE_CXX11_DELETED_FUNCTIONS)
+
+hpx_check_for_cxx11_explicit_cvt_ops(
+  DEFINITIONS HPX_HAVE_CXX11_EXPLICIT_CONVERSION_OPERATORS)
+
+hpx_check_for_cxx11_function_template_default_args(
+  REQUIRED "HPX needs support for C++11 defaulted function template arguments")
+
```

Comment:
Where do we rely on defaulted function template arguments? I thought we wouldn't...